toolkit/docs/how_it_works/3_package_building.md

```diff
@@ -181,42 +171,22 @@ Once the packages are cached they are copied into `./../out/RPMs` for use in fur
-
-A very simple workplan might look like:
-```makefile
-.PHONY: GOAL_PackagesToBuild
-GOAL_PackagesToBuild: ;
-GOAL_PackagesToBuild:  RUN_0_./../build/INTERMEDIATE_SRPMS/x86_64/example-1.0.0-1.cm1.src.rpm_example
-
-.PHONY: RUN_0_./../build/INTERMEDIATE_SRPMS/x86_64/example-1.0.0-1.cm1.src.rpm_example
-RUN_0_/home/damcilva/repos/demo/build/INTERMEDIATE_SRPMS/x86_64/example-1.0.0-1.cm1.src.rpm_example: ;
-RUN_0_/home/damcilva/repos/demo/build/INTERMEDIATE_SRPMS/x86_64/example-1.0.0-1.cm1.src.rpm_example:  BUILD_./../build/INTERMEDIATE_SRPMS/x86_64/example-1.0.0-1.cm1.src.rpm
-
-.PHONY: BUILD_./../build/INTERMEDIATE_SRPMS/x86_64/example-1.0.0-1.cm1.src.rpm
-BUILD_./../build/INTERMEDIATE_SRPMS/x86_64/example-1.0.0-1.cm1.src.rpm:
-	MAKEFLAGS= $(go-pkgworker) --input=./../build/INTERMEDIATE_SRPMS/x86_64/example-1.0.0-1.cm1.src.rpm --work-dir=$(CHROOT_DIR) ... >> $(LOGS_DIR)/pkggen/failures.txt 
-```
+`scheduler` controls a pool of build agents (`pkgworker`). It starts at the leaf nodes of the dependency graph and processes every node. Only processing `TypeBuild` nodes causes srpms to be built. Other node types are effectively NoOps, only processed to enforce dependency ordering.
 
-The entry point is `GOAL_PackagesToBuild`, which depends on the run node for the desired package (since run nodes indicate a package can be installed). Each run node will list its dependencies.
+`scheduler` will avoid building an srpm if it detects the package has already been built, and all of its dependencies were also prebuilt. If any dependencies of an SRPM needed to be built, then that SRPM will be built regardless.
 
-In this case the example package has no other dependencies, so it only depends on its own build node. That build node invokes the `pkgworker` tool to actually build the `*.rpm` file.
+`scheduler` supports dynamic dependencies. These are dependencies a package has on an implicit provide from another package. For example, package `foo` may `Requires: pkgconfig(bar)`. When `grapher` runs it is not known which package will provide `pkgconfig(bar)`. It is only known after packages are built and one of them reports that it provides `pkgconfig(bar)`. To handle this `scheduler` analyzes every rpm built for these implicit provides. If it finds one that is needed by another package in the graph it will modify the graph's nodes and edges so that it reflects this new information.
 
-### Stage 5: Pkgworker
-The `pkgworker` tool is not invoked directly by the build system, instead it is invoked from a recursive `Make` call to the dynamically generated `workplan.mk` file.
+During the build `scheduler` will continually try to optimize the build graph. When the graph is first loaded, `scheduler` walks the dependencies of every package that needs to be built. If it does not find any dynamic dependencies it will create a subgraph that only contains the needed packages.
```

Comment:
>subgraph [](start = 249, length = 8)

Out of curiosity: what is removed from the original one to form this subgraph?

Reply:
The scheduler creates a goal node of all explicitly desired packages (such as ones needed for image composition). Any nodes / packages that do not fall under the subgraph of this goal node would be removed when possible.